Rework post-build to support multiple executables

LDong-Arm opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

The current implementation of post-build operations always assumes there's only one CMake executable target, at the root of the build directory. When a project creates multiple executables under different subdirectories, the target-specific post-build command fails to locate each built executable to act on.

This PR reworks the post-build mechanism to allow running the command once per executable target with correct path to executable. For details, refer to the commit message of "_Rework post-build to support multiple executables_".

To verify support for multiple executables (and building apps with Mbed CLI 2 in general), this PR adds a test project containing two executables and a new GitHub Workflow to build it.

Fixes: #14933

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

A user project can define multiple executables, as long as each has `mbed_set_post_build()` called.

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->

None

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [x] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [x] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

@ARMmbed/mbed-os-core 

----------------------------------------------------------------------------------------------------------------

rwalton-arm committed:
CMake: Move post build functions out of root CMakeLists.txt

The 'post build' functions are made visible by adding the mbed-os
subdirectory. This is not ideal as any components in mbed-os wishing to
call the functions must be added after the functions are defined. To
improve modularity move these functions to a separate CMake script.

We include the post build CMake script in app.cmake for now so we don't
break user's projects.

LDong-Arm committed:
Rework post-build to support multiple executables

When building greentea tests, each test is an executable with its
own output binary path. This is also the case when a user project
produces multiple executables. But the current implementation of
post-build operations always assumes there's only one executable,
at the root of the build directory.

The post-build command depends on Mbed target, and it always takes
the the executable we build as an input file. To achieve this, we
let each Mbed target (that has a post-build command) define a function

    function(mbed_post_build_function target)

which takes a CMake executable target as an argument from which it can
get its binary path using generator expressions. It generates and adds
to the passed executable target a post-build custom command.

Notes:
* The function name needs to be exact, because CMake only supports
literal function calls - CMake can't dereference a function name from
a variable. To avoid multiple definitions of this function, each Mbed
target needs to guard it with a macro to check if the user is
building this Mbed target.
* `mbed_post_build_function()` is a function, but it is usually
defined by another macro rather than a parent function, because
nesting functions would make many variables inaccessible inside the
innermost `mbed_post_build_function()`.
* There's no more need to force regenerate images. Previously, post-
build commands were custom *targets* which always got to run, so we
force regenerated images on every build to avoid patching an image
that's already been patched once on previous build. Now post-build
commands are custom *commands* of the same executable target, and they
are only run if the executable target itself is rebuilt.

LDong-Arm committed:
CMake: Add test for multiple-executable support

Add a test to build two executables in two directories under a single
project.

LDong-Arm commented on targets/TARGET_Cypress/scripts/mbed_set_post_build_cypress.cmake:
```diff
@@ -3,72 +3,90 @@
-    endif()
+            if(${num_extra_args} GREATER 0)
+                # Get extra argument as `cortex_m0_hex`
+                list(GET extra_macro_args 0 cortex_m0_hex)
```
@rwalton-arm commented that the use of `${ARGN}` to get optional `--m0hex` image path is not rigorous. While it's a preexisting issue and not introduced by this PR, please let me know if we want to improve it here.

Patater replied:
Seems small enough to just fix rather than go through the overhead of filing and prioritizing an issue. Worth a separate commit in this PR to fix.

LDong-Arm replied:
I'll fix it

LDong-Arm replied:
I actually need some clarification, so made a reply to his original question: https://github.com/rwalton-arm/mbed-os/pull/1/files/eb17ce40d448eaee31fd6abd1963a11c439a05a5#r674938299

LDong-Arm replied:
Done, added a commit

Patater commented on .github/workflows/test_building_multiple_executables.yml:
```diff
@@ -0,0 +1,32 @@
+on:
+  pull_request:
+    paths:
+      - 'CMakeLists.txt'
```
Will this run on only the base CMakeLists.txt changing or do we need `**/CMakeLists.txt` to run on any CMakeLists.txt changes no matter the base folder?

LDong-Arm replied:
My initial idea was to only check places that are relevant to build/post-build. But now I start to think it doesn't hurt to run this check on every PR, as a general CMake check. Will remove the rules here.

rwalton-arm commented on tools/cmake/tests/multiple_executables/app1/CMakeLists.txt:
```diff
@@ -0,0 +1,10 @@
+# Copyright (c) 2021 Arm Limited. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+set(APP_TARGET app1)
```
Why define `APP_TARGET`? Seems like far less typing to just reuse `app1`, or `app2`. Is this part of some Mbed CMake style guide, or is mbed-os itself still relying on `APP_TARGET` somewhere?

LDong-Arm replied:
I'll change them to `app1`, `app2`.

rwalton-arm commented on CMakeLists.txt:
```diff
@@ -144,12 +144,16 @@ target_include_directories(mbed-core
 add_library(mbed-device_key INTERFACE)
 add_library(mbed-rtos INTERFACE)
 
+# Include targets/ first, because any post-build hook needs to be defined
```
In the commit message for commit "Rework post-build to support multiple executables", we say "guard it with a macro to **checks** if", that should be "guard it with a macro to **check** if"

LDong-Arm replied:
Good catch. I probably forgot to change it when rephrasing the sentence. 😅

LDong-Arm committed:
Cypress: Improve `mbed_post_build_psoc6_merge_hex()`

The CMake macro `mbed_post_build_psoc6_merge_hex()` takes the name of
a Cypress target and an optional Cortex-M0 hex image as arguments. The
proper way to define and parse optional arguments of a function or
macro is `cmake_parse_arguments()`, which allows the caller to
indicate what they are passing rather than rely on an argument's
relative position within `${ARGN}` which is not rigorous.

Also, avoid duplicating the common part of the post build command
when the optional argument is passed/not passed.

LDong-Arm reviewed (commented):
@Patater @rwalton-arm Thanks for reviewing, I've addressed your comments.

Patater reviewed (approved):
LGTM

rwalton-arm commented on targets/TARGET_Cypress/scripts/mbed_set_post_build_cypress.cmake:
```diff
@@ -6,34 +6,35 @@ include(mbed_set_post_build)
-                    merge
-                    --elf $<TARGET_FILE_DIR:${target}>/$<TARGET_FILE_BASE_NAME:${target}>.elf
-                    --m4hex $<TARGET_FILE_DIR:${target}>/$<TARGET_FILE_BASE_NAME:${target}>.hex
+            if(NOT "${CYPRESS_CORTEX_M0_HEX}" STREQUAL "")
```
Looks much better to me, the intent is now clear.

rwalton-arm reviewed (approved):
Looks great to me. Nice work!
